src/hooks/UseForm/useForm/types/public.ts

```diff
@@ -0,0 +1,395 @@
+
+/**
+ * The type of data that can be used in the `useForm` - the same as standard input types
+ *
+ * @export
+ * @interface FormData
+ * @typedef {FormData}
+ */
+export interface FormData {
+  [Key: ObjectKey]: string | string[] | number | boolean | Date
+}
+
+/**
+ * Options for configuring a form field when registering it
+ *
+ * @export
+ * @interface RegisterOptions
+ * @typedef {RegisterOptions}
+ * @template {FormData} TData
+ * @template {keyof TData} [TFieldName=keyof TData]
+ * @template {ObjectKey} [TRefPropsKey='ref']
+ * @template {boolean} [TIsRequired=false]
+ */
+export interface RegisterOptions<
+  TData extends FormData,
+  TDimensions extends number = 0,
+  TFieldName extends keyof TData = keyof TData,
+  TRefPropsKey extends ObjectKey = 'ref',
+  TIsRequired extends boolean = false,
+> {
+  /**
+   * Overrides the name of the `ref` prop to the registered input
+   *
+   * @type {?TRefPropsKey}
+   */
+  refName?: TRefPropsKey
+  /**
+   * If true, adds a validation that the field has a value, outputting an error if not
+   *
+   * @type {?TIsRequired}
+   */
+  isRequired?: TIsRequired
+  isRequiredErrorMessageOverride?: string
+
+  /**
+   * Adds a custom validation to the field.
+   * If `isRequired` is `true`, the field is guaranteed to have a value.
+   *
+   * @type {?(
+   *     field: TIsRequired extends true ? TData[TFieldName] : TData[TFieldName] | undefined | null,
+   *     fields: TIsRequired extends true ? Partial<TData> & Pick<TData, TFieldName> : Partial<TData>,
+   *     graph: IGraph<Partial<TData>, TDimensions>,
+   *   ) => string | null}
+   */
+  validate?: (
+    field: TIsRequired extends true ? TData[TFieldName] : TData[TFieldName] | undefined | null,
+    fields: TIsRequired extends true ? Partial<TData> & Pick<TData, TFieldName> : Partial<TData>,
+    graph: IGraph<Partial<TData>, TDimensions>,
+  ) => string | null
+
+  /**
+   * The coordinates of the registered field, if in a form with TDimensions > 0
+   *
+   * @type {?CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>}
+   */
+  coordinates?: CoordinatesOrNever<TDimensions, CoordinatesOfLength<TDimensions>>
+}
+
+/**
+ * The result of the `register` function including:
+ * - A ref of the type of the field's element with name RegisterOptions.refName ?? 'ref'
+ * - An `onChange` function for updating form data values
+ * - An `onFocus` function for marking fields as touched
+ * - An `onBlur` function for triggering validations
+ * @export
+ * @typedef {RegisterResult}
+ * @template {FieldElement} [TFieldElement=FieldElement]
+ * @template {RefProps<TFieldElement>} [TRefProps={
+ *     ref: React.Ref<TFieldElement>
+ *   }]
+ */
+export type RegisterResult<
+  TFieldElement extends FieldElement = FieldElement,
+  TRefProps extends RefProps<TFieldElement> = {
+    ref: React.Ref<TFieldElement>
+  },
+> = TRefProps & {
+  onChange: (event: unknown) => void
```

Comment:
I think this `event: unknown` can be typed with either `ChangeEvent` or `ChangeEventHandler`:

```suggestion
  onChange: React.ChangeEventHandler<TFieldElement>
```